modules/grandpa/src/lib.rs

```diff
@@ -473,6 +500,69 @@ pub mod pallet {
 		/// The `current_set_id` argument of the `submit_finality_proof_ex` doesn't match
 		/// the id of the current set, known to the pallet.
 		InvalidAuthoritySetId,
+		/// The submitter wanted free execution, but we can't fit more free transactions
+		/// to the block.
+		FreeHeadersLimitExceded,
+		/// The submitter wanted free execution, but the difference between best known and
+		/// bundled header numbers is below the `FreeHeadersInterval`.
+		BelowFreeHeaderInterval,
+	}
+
+	/// Called when new free header is imported.
+	pub fn on_free_header_imported<T: Config<I>, I: 'static>() {
+		FreeHeadersRemaining::<T, I>::mutate(|count| {
+			*count = match *count {
+				// never set to `None` here - the signed extension assumes that it is `None`
+				// outside of block execution - i.e. when transaction is validatied from
+				// the transaction pool
+				Some(count) => Some(count.saturating_sub(1)),
+				None => None,
+			}
+		});
+	}
+
+	/// Return true if we may refund transaction cost to the submitter. In other words,
+	/// this transaction is considered as common good deed w.r.t to pallet configuration.
+	fn may_refund_call_fee<T: Config<I>, I: 'static>(
```

Comment:
Nit: I think part of this logic is duplicated here and in the call_ext. Maybe we can deduplicate. Also `is_mandatory_header` is present in the `submit_finality_proof_info_from_args`. Not sure if we should also send it as a parameter here.

Reply:
I've been considering that. But there are several tiny code differences, which turn into large behavior differences. I had two choices - either make one giant function to check everything and let it accept several params like `allow_none_free_remaining_headers`, `ignore_is_free_execution_expected`, ... or have a two separate functions. I've decided to go the latter way. If there's a consensus - I could change

Reply:
I was thinking this might be the case.

Anyway, can you create an issue for https://github.com/paritytech/parity-bridges-common/pull/2873#discussion_r1560909073 please and maybe mention this discussion also ? Maybe some small things can be done. Also you can assign it to me if you want. Or however you prefer.